share/wake/lib/core/json.wake

```diff
@@ -154,18 +154,155 @@ export def x // y =
     def helper (Pair (Pair k c) (Pair l d)) = k ==~ l && c ==/ d
-    if a.len != b.len then False else zip a b | forall helper
+    if a.len != b.len then
+      False
+    else
+      zip a b
+      | forall helper
   (JArray   a) (JArray   b) =
     def helper (Pair c d) = c ==/ d
     if a.len != b.len then False else zip a b | forall helper
   (JArray a) _ = exists (_ ==/ y) a
   _ (JArray b) = exists (_ ==/ x) b
   _ _ = False
+
+# Simplify a JSON value according to the rules given.
+# For example, Wake's implementation of JSON is slightly more lenient than the
+# base standards, so if maximum compatibility is desired in situations where
+# some other implementation may be parsing generated output,
+# `normalizeJSONBasic` may be used to target that least common denominator.
+export def normalizeJSON (fmt: JSONNormalize): JValue => Result JValue Error = match _
+    JString a =
+        getJSONNormalizeString fmt a
+        | rmap JString
+    JInteger a =
+        getJSONNormalizeInteger fmt a
+        | rmap JInteger
+    JDouble a =
+        getJSONNormalizeDouble fmt a
+        | rmap JDouble
+    JBoolean a =
+        getJSONNormalizeBoolean fmt a
+        | rmap JBoolean
+    JNull =
+        Pass JNull
+    JObject a =
+        getJSONNormalizeObject fmt a
+        | rmap JObject
+    JArray a =
+        getJSONNormalizeArray fmt a
+        | rmap JArray
+
+# The rules by which `normalizeJSON` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONDefault: JSONNormalize =
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONBasic: JSONNormalize =
+    def nfkcKeys =
+        map (_ | editPairFirst unicodeIdentifier)
+    normalizeJSONDefault
+    | setJSONNormalizeString (Pass _.unicodeCanonical)
+    | setJSONNormalizeDouble filterNonDigitJDouble
+    | setJSONNormalizeObject (deduplicateJObjectKeys _.nfkcKeys)
+
+# Fail on any `Double` values which can't be represented numerically.
+# Both JSON specifications describe their double values as allowing decimal or
+# exponential forms, but don't implement the full IEEE standard.
+def filterNonDigitJDouble (n: Double): Result Double Error = match n.dclass
+    DoubleInfinite =
+        failWithError "For compatibility, JSON doubles should not be infinite."
```

Comment:
It would be pretty useful to indicate the 'path' to the object where this error occurred in the message. (eg: `Fail "document/jingle-bells/pages: infinite Double forbidden in json <= json5"`)